File: implementations/python/src/esque/esque.py
from __future__ import annotations

import logging
import time
from datetime import UTC, datetime
from types import TracebackType

from elasticsearch import Elasticsearch

from esque.configuration import EsqueConfiguration
from esque.elasticsearch.documents import MigrationRecord
from esque.elasticsearch.lock import ElasticsearchDocumentLock
from esque.elasticsearch.operations import RestClientOperations
from esque.migration.loader import MigrationFileLoader
from esque.migration.model import MigrationFile
from esque.migration.template import MigrationTemplateResolver

log = logging.getLogger(__name__)


class Esque:
    def __init__(
        self,
        client: Elasticsearch,
        configuration: EsqueConfiguration,
        properties: dict[str, str] | None = None,
    ) -> None:
        self._configuration = configuration
        self._migration_loader = MigrationFileLoader(
            configuration.migration_directory,
            MigrationTemplateResolver(properties or {}),
        )
        self._operations = RestClientOperations(client, configuration.migration_key)
        self._lock = ElasticsearchDocumentLock(self._operations)

    def __enter__(self) -> Esque:
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None:
        self.close()

    def close(self) -> None:
        try:
            self._lock.unlock()
        except RuntimeError:
            pass  # lock was not held — expected at end of a clean run
        except Exception:
            log.warning(
                "failed to release a execution lock. you may need to manually delete the lock document yourself"
            )

        try:
            self._operations.close()
        except Exception:
            log.warning("failed to close client. this is likely not an issue")

    def execute(self) -> None:
        log.info("Starting esque execution")
        try:
            self._initialize()
            files = self._migration_loader.load()
            history = self._operations.get_migration_records()
            self._verify_state_integrity(files, history)
            self._run_migrations(files)
            log.info("Completed esque execution")
        except Exception as e:
            raise RuntimeError("Failed to run esque execution") from e

    def _initialize(self) -> None:
        log.info("Initializing esque as needed")
        if not self._operations.check_migration_index_exists():
            self._operations.create_migration_index()

    def _verify_state_integrity(self, files: list[MigrationFile], history: list[MigrationRecord]) -> None:
        log.info("Verifying integrity of migration state as compared to found migration files")
        if len(history) > len(files):
            raise RuntimeError(
                "the migration records are showing more migrations than the local system defines. "
                "did you refactor your files or use an incorrect migration key?"
            )
        if history and len(history) != history[-1].order + 1:
            raise RuntimeError("the migration records seem to be corrupt as some records appear to be missing.")
        for record in history:
            self._verify_record_integrity(record, files)
        log.info("Integrity checks passed")

    def _verify_record_integrity(self, record: MigrationRecord, files: list[MigrationFile]) -> None:
        companion = next((f for f in files if f.metadata.filename == record.filename), None)
        if companion is None:
            raise RuntimeError(
                f"could not find migration file matching migration history record by filename [{record.filename}]"
            )
        if (
            record.order != files.index(companion)
            or record.version != companion.metadata.version
            or record.description != companion.metadata.description
            or record.checksum != companion.metadata.checksum
            or record.migration_key != self._configuration.migration_key
        ):
            raise RuntimeError(
                f"could not verify integrity of migration history record for filename [{record.filename}]. "
                "did you refactor your migration scripts after a previous execution?"
            )

    def _run_migrations(self, files: list[MigrationFile]) -> None:
        try:
            for file in files:
                try:
                    log.info("Attempting to acquire lock for execution")
                    if self._lock.try_lock(self._configuration.lock_timeout_minutes):
                        log.info("Lock acquired. Executing queries in migration file [%s]", file.metadata.filename)
                        if self._operations.get_migration_record_for_migration_file(file) is not None:
                            log.info(
                                "Migration for file [%s] and key [%s] already executed. Skipping",
                                file.metadata.filename,
                                self._configuration.migration_key,
                            )
                        else:
                            start = time.monotonic()
                            self._run_migration_for_file(file)
                            elapsed_ms = int((time.monotonic() - start) * 1000)
                            log.info(
                                "Execution complete for migration file [%s]. Took [%d] milliseconds",
                                file.metadata.filename,
                                elapsed_ms,
                            )
                            self._operations.create_migration_record(
                                MigrationRecord(
                                    migration_key=self._configuration.migration_key,
                                    order=files.index(file),
                                    filename=file.metadata.filename,
                                    version=file.metadata.version,
                                    description=file.metadata.description,
                                    checksum=file.metadata.checksum,
                                    installed_by=self._configuration.migration_user,
                                    installed_on=datetime.now(UTC),
                                    execution_time=elapsed_ms,
                                )
                            )
                    else:
                        log.error("Failed to acquire lock in the allotted time. Did a lock not get cleared?")
                        raise RuntimeError("failed to acquire lock")
                except Exception as e:
                    raise RuntimeError(f"Failed to execute queries in migration file [{file.metadata.filename}]") from e
                finally:
                    log.info("Releasing execution lock")
                    self._lock.unlock()
        except Exception as e:
            raise RuntimeError("failed to run migrations") from e

    def _run_migration_for_file(self, file: MigrationFile) -> None:
        log.info("Executing queries defined in migration file [%s]", file.metadata.filename)
        for position, definition in enumerate(file.contents.requests):
            try:
                log.info(
                    "Executing query in position [%d] in migration file [%s]",
                    position,
                    file.metadata.filename,
                )
                self._operations.execute_migration_definition(definition)
                log.info(
                    "Query in position [%d] in migration file [%s] executed successfully",
                    position,
                    file.metadata.filename,
                )
            except Exception as e:
                raise RuntimeError(
                    f"Failed to execute query in position [{position}] in migration file [{file.metadata.filename}]"
                ) from e
        log.info("Execution complete for queries in migration file [%s]", file.metadata.filename)

File: implementations/typescript/src/esque.ts
import type { Client } from "@elastic/elasticsearch";
import type { EsqueConfiguration } from "./configuration.js";
import type { MigrationRecord } from "./elasticsearch/documents.js";
import { ElasticsearchDocumentLock, LockNotHeldError } from "./elasticsearch/lock.js";
import { RestClientOperations } from "./elasticsearch/operations.js";
import { MigrationFileLoader } from "./migration/loader.js";
import type { MigrationFile } from "./migration/model.js";
import { MigrationTemplateResolver } from "./migration/template.js";

export class Esque {
  private readonly configuration: EsqueConfiguration;
  private readonly migrationLoader: MigrationFileLoader;
  private readonly operations: RestClientOperations;
  private readonly lock: ElasticsearchDocumentLock;

  constructor(client: Client, configuration: EsqueConfiguration, properties: Record<string, string> = {}) {
    this.configuration = configuration;
    this.migrationLoader = new MigrationFileLoader(
      configuration.migrationDirectory,
      new MigrationTemplateResolver(properties),
    );
    this.operations = new RestClientOperations(client, configuration.migrationKey);
    this.lock = new ElasticsearchDocumentLock(this.operations);
  }

  async close(): Promise<void> {
    try {
      await this.lock.unlock();
    } catch (error) {
      if (!(error instanceof LockNotHeldError)) {
        console.warn("failed to release a execution lock. you may need to manually delete the lock document yourself");
      }
    }
    try {
      await this.operations.close();
    } catch {
      console.warn("failed to close client. this is likely not an issue");
    }
  }

  async [Symbol.asyncDispose](): Promise<void> {
    await this.close();
  }

  async execute(): Promise<void> {
    try {
      await this.initialize();
      const files = this.migrationLoader.load();
      const history = await this.operations.getMigrationRecords();
      this.verifyStateIntegrity(files, history);
      await this.runMigrations(files);
    } catch (error) {
      throw new Error("Failed to run esque execution", { cause: error });
    }
  }

  private async initialize(): Promise<void> {
    if (!(await this.operations.checkMigrationIndexExists())) {
      await this.operations.createMigrationIndex();
    }
  }

  private verifyStateIntegrity(files: MigrationFile[], history: MigrationRecord[]): void {
    if (history.length > files.length) {
      throw new Error(
        "the migration records are showing more migrations than the local system defines. " +
          "did you refactor your files or use an incorrect migration key?",
      );
    }
    const last = history[history.length - 1];
    if (last !== undefined && history.length !== last.order + 1) {
      throw new Error("the migration records seem to be corrupt as some records appear to be missing.");
    }
    for (const record of history) {
      this.verifyRecordIntegrity(record, files);
    }
  }

  private verifyRecordIntegrity(record: MigrationRecord, files: MigrationFile[]): void {
    const companion = files.find((f) => f.metadata.filename === record.filename);
    if (companion === undefined) {
      throw new Error(
        `could not find migration file matching migration history record by filename [${record.filename}]`,
      );
    }
    if (
      record.order !== files.indexOf(companion) ||
      record.version !== companion.metadata.version ||
      record.description !== companion.metadata.description ||
      record.checksum !== companion.metadata.checksum ||
      record.migrationKey !== this.configuration.migrationKey
    ) {
      throw new Error(
        `could not verify integrity of migration history record for filename [${record.filename}]. ` +
          "did you refactor your migration scripts after a previous execution?",
      );
    }
  }

  private async runMigrations(files: MigrationFile[]): Promise<void> {
    try {
      for (const file of files) {
        let executionError: unknown;
        try {
          if (await this.lock.tryLock(this.configuration.lockTimeoutMinutes)) {
            const existing = await this.operations.getMigrationRecordForMigrationFile(file);
            if (existing === null) {
              const start = performance.now();
              await this.runMigrationForFile(file);
              const elapsedMs = Math.round(performance.now() - start);
              await this.operations.createMigrationRecord({
                migrationKey: this.configuration.migrationKey,
                order: files.indexOf(file),
                filename: file.metadata.filename,
                version: file.metadata.version,
                description: file.metadata.description,
                checksum: file.metadata.checksum,
                installedBy: this.configuration.migrationUser,
                installedOn: new Date().toISOString(),
                executionTime: elapsedMs,
              });
            }
          } else {
            throw new Error("failed to acquire lock");
          }
        } catch (error) {
          executionError = new Error(`Failed to execute queries in migration file [${file.metadata.filename}]`, {
            cause: error,
          });
        }

        // Never let a release failure silently replace an execution failure (or its success) —
        // capture both separately instead of releasing in a `finally` that could throw over
        // whatever the try/catch above was about to produce.
        let releaseError: unknown;
        try {
          await this.lock.unlock();
        } catch (error) {
          if (!(error instanceof LockNotHeldError)) {
            releaseError = error;
          }
        }

        if (executionError !== undefined) {
          if (releaseError !== undefined) {
            console.warn(
              `failed to release execution lock for migration file [${file.metadata.filename}] after a migration failure. you may need to manually delete the lock document yourself`,
            );
          }
          throw executionError;
        }
        if (releaseError !== undefined) {
          throw new Error(`Failed to release execution lock after migration file [${file.metadata.filename}]`, {
            cause: releaseError,
          });
        }
      }
    } catch (error) {
      throw new Error("failed to run migrations", { cause: error });
    }
  }

  private async runMigrationForFile(file: MigrationFile): Promise<void> {
    for (const [position, definition] of file.contents.requests.entries()) {
      try {
        await this.operations.executeMigrationDefinition(definition);
      } catch (error) {
        throw new Error(
          `Failed to execute query in position [${position}] in migration file [${file.metadata.filename}]`,
          { cause: error },
        );
      }
    }
  }
}
